### PR TITLE
[flang][openacc] Do not loose attributes on folding

### PR DESCRIPTION
**Pull request by clementval**
hlfir.declare introduce some boxes that can be later optimized away. The OpenACC lowering is currently setting some attributes on FIR operations to track declare variables. When the boxes are optimized away these attributes are lost. This patch propagate OpenACC attributes from box_addr op to the defining op of the folding result.

### CODE REVIEW

File: flang/lib/Optimizer/Dialect/FIROps.cpp [hunk 1]
```diff
@@ -36,6 +37,18 @@ namespace {
 #include "flang/Optimizer/Dialect/CanonicalizationPatterns.inc"
 } // namespace
 
+static void propagateAttributes(mlir::Operation *fromOp,
+                                mlir::Operation *toOp) {
+  if (!fromOp || !toOp)
+    return;
+
+  for (mlir::NamedAttribute attr : fromOp->getAttrs()) {
+    if (attr.getName().str().rfind(
```
Comment (vzakhari) [inline flang/lib/Optimizer/Dialect/FIROps.cpp:46]: nit: can we use `starts_with` here to make it more readable?

File: flang/test/Fir/OpenACC/propagate-attr-folding.fir [hunk 2]
```diff
@@ -0,0 +1,39 @@
+// RUN: fir-opt %s --opt-bufferization | FileCheck %s 
+
+// Check that OpenACC attributes are propagated to the defining operations when
+// fir.box_addr is folded in bufferization optimization.
```
Comment (vzakhari) [inline flang/test/Fir/OpenACC/propagate-attr-folding.fir:4]: Did you intend to use `fir.box_addr` in the test?  I do not see any :)

Comment (clementval) [reply flang/test/Fir/OpenACC/propagate-attr-folding.fir:4]: My bad I pushed the wrong test. I'll update.

File: flang/lib/Optimizer/Dialect/FIROps.cpp (hunk 1 above)
Comment (clementval) [reply flang/lib/Optimizer/Dialect/FIROps.cpp:46]: I might have overlooked it but std::string does not have `start_with`?

**Review by vzakhari — APPROVED** [review]
Thank you, Valentin!

**Pull request merged**